### ISSUE TITLE
Upgrade planned even with diff errors

### ISSUE THREAD

**Issue by mickeder** [Praqma/helmsman#562]
Hi @luisdavim,
Your last change creates an issue when diff returns error, upgrade is still planned.
https://github.com/Praqma/helmsman/blob/9e609e3830b6c05b8be932f84eec4393a50ac337/internal/app/decision_maker.go#L327-L338

Probably `return err` instead of `log.Error` in L330 should resolve the issue.

**Comment by luisdavim** [comment]
that was on purpose, in this case either `--always-upgrade` is set or the chart has changed so we don't care about the diff, that is only used that in case `--verbose` or `--show-diff` are set but that would be for visibility only.

**Comment by mickeder** [comment]
Ok, so the issue is not related to the latest changes, but earlier with the introduction of this line
https://github.com/Praqma/helmsman/blob/bd4765591f7c36d235c5c666d067fc4abab7cfa0/internal/app/decision_maker.go#L323-L327
instead of
https://github.com/Praqma/helmsman/blob/aad626389aa3943b333ae38423dbb0b30f835692/internal/app/decision_maker.go#L397-L402

Earlier the diffing was performed when the version of chart changed, so when the new version caused diff errors, helmsman returned error even without `--apply`. I raised this issue, cause in our pipeline first we check if everything is ok by running helmsman without `--apply`, so when any chart causes diff error, this results in failed pipeline. In latest version, we could get an error only after setting `--apply`.

I don't think it's redundant to diff when `--always-upgrade` is set, the chart has changed or the chart version changed, because diffing could return error even without setting `--apply`.

**Comment by luisdavim** [comment]
Ok, what sort of diff errors would you catch? I thought `helm diff` was only returning errors if it would fail to calculate the diff.

I'll look into adding the diff when `--apply` is not set.

**Comment by mickeder** [comment]
`helm diff` also returns validation errors of charts, e.g.
`Failed to render chart: exit status 1: Error: unable to build kubernetes objects from release manifest: 
error validating "": error validating data: ValidationError(Deployment.spec): unknown field "updateStrategy" 
in io.k8s.api.apps.v1.DeploymentSpec`

**Comment by luisdavim** [comment]
I've been thinking about this, the changes you're referring to were made to avoid running the diff command when not needed for the decision making process and the fact that `helm diff` validates the chart was more of a side effect than anything else.

What if I add a condition that would run the diff command if `--apply` is not set as a way to validate the chart?

Another question this brings to mind is what is the purpose of having both `--apply` and `--dry-run` flags? I think that if you run your CI pipeline with `--dry-run` you'll get the effect you're looking for.
Is there really a need to maintain both these flags? should we just drop the `--apply` and use the `--dry-run` for the cases when `helmsman` is ran without the `--apply` flag? Maybe we should turn the `--apply` and `--destroy` into sub-commands to make this more explicit and avoid unwanted side effects when people upgrade to the version implementing this change.

CC: @mkubaczyk @sami-alajrami @jmontroy90 @antoinedeschenes what do you guys think?

**Comment by sami-alajrami** [comment]
I agree on the subcommands idea for `apply` and `destroy`.
As for `--dry-run` , I recall that it was added to allow executing helmsman helm commands with the dry-run flag and log the output. If we go with subcommands, we can have a plan subcommand where the `--dry-run` flag can be used. It does not make sense to use it with `apply`.

As for the original issue here, I think something like `helmsman plan --dry-run` command would catch the chart validation error mentioned as the helm upgrade `--dry-run` command should return the same error.

**Comment by antoinedeschenes** [comment]
I suppose these errors would be caught running helmsman with `--apply --dry-run`? (Since it's a rendering error, `helm template` would be more efficient than `helm diff` at finding these errors)

As we're only running a "plan" by default, the current `--dry-run` alone doesn't seem to make sense to me (it only makes sense to use that flag with helm install, upgrade, uninstall, rollback)
